File: scripts/ncm/generate-ncm-embeddings.mjs
/**
 * Gera ncm-embeddings.json usando @xenova/transformers (mesmo modelo do browser).
 * Uso: npm install && npm run generate-embeddings
 */

import { pipeline } from '@xenova/transformers';
import { readFileSync, writeFileSync, mkdirSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT = join(__dirname, '..', '..');
const TABLE_PATH = join(ROOT, 'docs', 'NCM', 'Tabela_NCM.json');
const OUT_PATH = join(ROOT, 'docs', 'NCM', 'ncm-embeddings.json');

const MODEL_ID = 'Xenova/paraphrase-multilingual-MiniLM-L12-v2';

function flattenNcmTable(data) {
  const out = [];
  if (!data || typeof data !== 'object') return out;
  for (const levelKey of Object.keys(data)) {
    const root = data[levelKey];
    if (!root?.capitulos) continue;
    for (const capKey of Object.keys(root.capitulos)) {
      const cap = root.capitulos[capKey];
      const ncms = cap?.ncms || {};
      for (const ncmKey of Object.keys(ncms)) {
        const n = ncms[ncmKey];
        if (!n?.codigo || !n?.descricao) continue;
        const c = String(n.codigo).replace(/\D/g, '');
        if (c.length !== 8) continue;
        const c4 = c.slice(0, 4);
        const c6 = c.slice(0, 6);
        const d4 = (ncms[c4]?.descricao || '').trim();
        const d6 = (ncms[c6]?.descricao || '').trim();
        const full = [d4, d6, n.descricao].filter(Boolean).join(' ').replace(/\s+/g, ' ').trim();
        out.push({
          codigo: c,
          descricao: n.descricao,
          descricao4: d4,
          descricao6: d6,
          descricaoCompleta: full || n.descricao,
          capitulo: String(capKey).replace(/\D/g, '').slice(0, 2),
        });
      }
    }
  }
  return out;
}

async function main() {
  console.log('Carregando Tabela_NCM.json...');
  const data = JSON.parse(readFileSync(TABLE_PATH, 'utf8'));
  const items = flattenNcmTable(data);
  console.log('NCMs 8 dígitos:', items.length);

  console.log('Carregando modelo', MODEL_ID, '...');
  const extractor = await pipeline('feature-extraction', MODEL_ID, {
    pooling: 'mean',
    normalize: true,
  });

  const BATCH = 64;
  const out = [];
  for (let i = 0; i < items.length; i += BATCH) {
    const batch = items.slice(i, i + BATCH);
    const texts = batch.map((x) => x.descricaoCompleta);
    const t = await extractor(texts, { pooling: 'mean', normalize: true });
    const dim = t.dims[t.dims.length - 1];
    const arr = Array.from(t.data);
    for (let j = 0; j < batch.length; j++) {
      const row = batch[j];
      const emb = arr.slice(j * dim, (j + 1) * dim).map((x) => Math.round(Number(x) * 1e6) / 1e6);
      out.push({
        codigo: row.codigo,
        descricao: row.descricao,
        descricao4: row.descricao4,
        descricao6: row.descricao6,
        capitulo: row.capitulo,
        embedding: emb,
      });
    }
    if ((i + BATCH) % 500 === 0 || i + BATCH >= items.length) {
      console.log('Embeddings:', Math.min(i + BATCH, items.length), '/', items.length);
    }
  }

  mkdirSync(join(ROOT, 'docs', 'NCM'), { recursive: true });
  writeFileSync(OUT_PATH, JSON.stringify(out), 'utf8');
  console.log('Salvo em', OUT_PATH);
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: scripts/ncm/generate_ncm_embeddings.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Gera embeddings para cada NCM de 8 dígitos a partir da descrição completa (4+6+8).
Saída: ncm-embeddings.json para uso com ncm-embeddings.js (Transformers.js).
Modelo: paraphrase-multilingual-MiniLM-L12-v2 (mesmo usável no browser).
"""

import json
import sys
from pathlib import Path

# Caminhos (ajustar se necessário)
ROOT = Path(__file__).resolve().parents[2]
TABLE_PATH = ROOT / "docs" / "NCM" / "Tabela_NCM.json"
OUT_PATH = ROOT / "docs" / "NCM" / "ncm-embeddings.json"


def flatten_ncm_table(data):
    """Réplica a lógica do ncm-motor: só 8 dígitos, desc4+desc6+desc8."""
    out = []
    if not data or not isinstance(data, dict):
        return out
    for level_key, root in data.items():
        if not root or not isinstance(root.get("capitulos"), dict):
            continue
        for cap_key, cap in root["capitulos"].items():
            ncms = cap.get("ncms") or {}
            for ncm_key, n in ncms.items():
                if not n or not n.get("codigo") or not n.get("descricao"):
                    continue
                c = "".join(ch for ch in str(n["codigo"]) if ch.isdigit())
                if len(c) != 8:
                    continue
                c4, c6 = c[:4], c[:6]
                d4 = (ncms.get(c4) or {}).get("descricao") or ""
                d6 = (ncms.get(c6) or {}).get("descricao") or ""
                parts = [d4, d6, n["descricao"]]
                full = " ".join(p for p in parts if p).replace("\n", " ").strip()
                full = " ".join(full.split())
                out.append({
                    "codigo": c,
                    "descricao": n["descricao"],
                    "descricao4": d4,
                    "descricao6": d6,
                    "descricaoCompleta": full or n["descricao"],
                    "capitulo": "".join(ch for ch in str(cap_key) if ch.isdigit())[:2],
                })
    return out


def main():
    print("Carregando Tabela_NCM.json...")
    with open(TABLE_PATH, "r", encoding="utf-8") as f:
        data = json.load(f)

    items = flatten_ncm_table(data)
    print(f"NCMs de 8 dígitos: {len(items)}")

    print("Carregando modelo sentence-transformers...")
    from sentence_transformers import SentenceTransformer
    model = SentenceTransformer("paraphrase-multilingual-MiniLM-L12-v2")

    texts = [x["descricaoCompleta"] for x in items]
    print("Gerando embeddings...")
    embs = model.encode(texts, show_progress_bar=True, normalize_embeddings=True)

    out = []
    for i, row in enumerate(items):
        out.append({
            "codigo": row["codigo"],
            "descricao": row["descricao"],
            "descricao4": row["descricao4"],
            "descricao6": row["descricao6"],
            "capitulo": row["capitulo"],
            "embedding": [round(float(x), 6) for x in embs[i].tolist()],
        })

    OUT_PATH.parent.mkdir(parents=True, exist_ok=True)
    with open(OUT_PATH, "w", encoding="utf-8") as f:
        json.dump(out, f, ensure_ascii=False, separators=(",", ":"))

    print(f"Salvo em {OUT_PATH}")


if __name__ == "__main__":
    main()
